File: src/modules/technique/domain/technique.entity.ts
import { TrainingSession } from "src/modules/training-session/domain/training-session.entity";
import { Column, Entity, ManyToMany, PrimaryGeneratedColumn } from "typeorm";

@Entity()
export class Technique {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ unique: true })
  name: string;

  @Column({ nullable: false })
  description: string;

  @Column({ type: 'enum', enum: ['SUBMISSION', 'SWEEP', 'TAKEDOWN', 'GUARD'] })
  type: string;

  @Column({ type: 'enum', enum: ['BASIC', 'INTERMEDIATE', 'ADVANCED'], default: 'BASIC' })
  difficulty: string;

  @ManyToMany(() => TrainingSession, (session) => session.techniques)
  trainingSession: TrainingSession[];
}

File: src/modules/technique/domain/technique_entity.go
package domain_technique

type Difficulty int

const (
	Beginner Difficulty = iota
	Intermediate
	Advanced
)

type Category int

const (
	Submission Category = iota
	Sweep
	Pass
	Guard
	Takedown
	Defend
	SubmissionEscape
)

type Technique struct {
	ID                    string     `json:"id" gorm:"primaryKey"`
	Name                  string     `json:"name" gorm:"not null, unique" `
	NamePortuguese        string     `json:"name_portuguese" gorm:"not null"`
	Description           string     `json:"description" gorm:"not null"`
	DescriptionPortuguese string     `json:"description_portuguese" gorm:"not null"`
	Category              Category   `json:"category" gorm:"not null"`
	Difficulty            Difficulty `json:"difficulty" gorm:"not null"`
}
